electricitymap/contrib/config/zones.py

```diff
@@ -0,0 +1,66 @@
+"""This module contains functions for working with zones configs."""
+
+from collections import defaultdict
+from copy import deepcopy
+from pathlib import Path
+from typing import Any, Dict, List
+
+from electricitymap.contrib.config.config_types import BoundingBox, ZoneKey
+
+
+def zone_bounding_boxes(zones_config: Dict[ZoneKey, Any]) -> Dict[ZoneKey, BoundingBox]:
+    """Returns a dict mapping each zone to its bounding box."""
+    bounding_boxes = {}
+    for zone_id, zone_config in zones_config.items():
+        if "bounding_box" in zone_config:
+            bounding_boxes[zone_id] = zone_config["bounding_box"]
+    return bounding_boxes
+
+
+def zone_parents(zones_config: Dict[ZoneKey, Any]) -> Dict[ZoneKey, ZoneKey]:
+    """Returns a dict mapping each zone to its parent zone."""
+    zone_parents = {}
+    for zone_id, zone_config in zones_config.items():
+        if "subZoneNames" in zone_config:
+            for sub_zone_id in zone_config["subZoneNames"]:
+                zone_parents[sub_zone_id] = zone_id
+    return zone_parents
+
+
+def generate_zone_neighbours(
+    zones_config: Dict[ZoneKey, Any], exchanges_config: Dict[str, Any]
+) -> Dict[ZoneKey, List[ZoneKey]]:
+    """Returns a dict mapping each zone to its neighbours.
+
+    Neighbours are defined as zones that are connected by an exchange.
+    This represents the edges of the flow-tracing graph.
+    """
+    zone_neighbours = defaultdict(set)
+    for k, v in exchanges_config.items():
+        if not v.get("parsers", {}).get("exchange", None):
+            # Interconnector config has no parser, and will therefore not be part
+            # of the flowtracing graph.
+            continue
+        zone_1, zone_2 = k.split("->")
+        if zones_config[zone_1].get("subZoneNames") or zones_config[zone_2].get(
+            "subZoneNames"
+        ):
```

Comment:
Here is a type error as we can't use the zone_1 and zone_2 strings to access the zone_config that has the type ZoneKey. We need to cast these to correct types first.

Reply:
I have a couple questions about this:

- Should I cast by just changing `zone_1` to  `ZoneKey(zone_1)` or using `typing.cast` (https://docs.python.org/3/library/typing.html#typing.cast) and use `cast(zone_1, ZoneKey)`?
- Is there anything I should run to check types? (The tests pass with no warnings. In general, the Python3 interpreter ignores type hints. I've used `mypy` to check type hints before... and I see that `pydantic` is used in config/model.py, which might do type checking?)

Reply:
We have no set typechecking in the CI yet and this is definitely something we should consider in the future. Personally I am using the Pylance VS code extension powered by https://github.com/microsoft/pyright

As for casting I think you can just do ZoneKey(zone_1).